### PR TITLE
Match expression

### LINKED ISSUE #48: Match expression in concrete and core
This should be very simple

### PR DESCRIPTION
**Pull request by wsx-ucb**
Fixes #48.
Current todo list:
- [x] Add match expression in concrete (and syntax?).
- [x] Correctly handle `Match` in various visitors (`Renamer`, `ExprTycker`, etc.).
- [ ] Unify the logic for pattern matching in function definition and that with match expression.
Main difference here is that we can simultaneously match multiple arguments in function definition, but only allowing one scrutinee in the match expression. Maybe just elaborate matching multiple arguments into matching a tuple.
- [x] Add test cases.

### CODE REVIEW

**Comment by ice1000** [comment]
Syntax is already there. Just patch the producer. However, I recommend you to wait until we migrate the frontend.

**Comment by ice1000** [comment]
:fire:

**Comment by ice1000** [comment]
ping

**Comment by wsx-ucb** [comment]
Type checking for `match` is implemented by reusing the `elabClausesClassified`.
However, currently not even the simplest case would pass, due to some weird patterns being generated in the match clauses.
I need to reinvestigate all the visitors used after `AyaGKProducer` for this matter.

**Comment by ice1000** [comment]
> Type checking for `match` is implemented by reusing the `elabClausesClassified`.
> However, currently not even the simplest case would pass, due to some weird patterns being generated in the match clauses.
> I need to reinvestigate all the visitors used after `AyaGKProducer` for this matter.

When do you have time? We can debug it together.

**Comment by wsx-ucb** [comment]
I've looked into this issue yesterday, and one problem is that `ExprTraversal` does not visit patterns.
This makes the pattern preprocessing logic applies to function declaration with pattern matching statements, but not to `match` expressions in function body.
The fix wouldn't be hard, but more likely there are other issues hiding around.

**Comment by ice1000** [comment]
> I've looked into this issue yesterday, and one problem is that `ExprTraversal` does not visit patterns.
> This makes the pattern preprocessing logic applies to function declaration with pattern matching statements, but not to `match` expressions in function body.
> The fix wouldn't be hard, but more likely there are other issues hiding around.

I think there is a subinterface of `ExprTraversal` that visits patterns. You should just copy and paste it.

**Comment by ice1000** [comment]
@wsx-ucb from now on, you no longer need to regenerate the lexer. Only the parser. Yay!

File: base/src/main/java/org/aya/concrete/desugar/Desugarer.java [hunk 1]
```diff
@@ -37,14 +37,14 @@ public static class DesugarInterruption extends Exception {}
 
   @Override public @NotNull Expr pre(@NotNull Expr expr) {
     return switch (expr) {
-      case Expr.AppExpr(var pos, Expr.RawSortExpr(var uPos, var kind), var arg)when kind == SortKind.Type -> {
+      case Expr.AppExpr(var pos,Expr.RawSortExpr(var uPos,var kind),var arg)when kind == SortKind.Type -> {
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/desugar/Desugarer.java:40]: Please upgrade IntelliJ

**Review by ice1000 — CHANGES_REQUESTED** [review]

File: base/src/main/java/org/aya/concrete/desugar/Desugarer.java (hunk 1 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/concrete/desugar/Desugarer.java:40]: I'm on 2022.2.3 now, but it seems that the formatting is still wrong.
Which version should I use then?

Comment (ice1000) [reply base/src/main/java/org/aya/concrete/desugar/Desugarer.java:40]: EAP

Comment (wsx-ucb) [reply base/src/main/java/org/aya/concrete/desugar/Desugarer.java]: Ok, I've switched to EAP and some spaces are added back.
But there is still no space between `)` and `when`, is this fine at this point?

**Comment by ice1000** [comment]
We should have a meeting and discuss these changes.

File: base/src/main/java/org/aya/concrete/desugar/Desugarer.java (hunk 1 above)
Comment (ice1000) [reply base/src/main/java/org/aya/concrete/desugar/Desugarer.java]: Yes

**Review by ice1000 — APPROVED** [review]
bors r+

**Pull request merged**